### PR TITLE
Retry connection to bootstrap nodes in discovery after failed attempt

### PR DESCRIPTION
**Pull request by casiojapi**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Requires Activation Code (Hard Fork)

* **Other information**:

### CODE REVIEW

**Review by Vovchyk — CHANGES_REQUESTED** [review]

File: rskj-core/src/main/java/co/rsk/net/discovery/PeerExplorer.java [hunk 1]
```diff
@@ -427,12 +429,25 @@ synchronized void update() {
...
+            this.checkPeersPulse(closestNodes);
+        }
+    }
 
-        this.askForMoreNodes(closestNodes);
-        this.checkPeersPulse(closestNodes);
+    private boolean shouldRetryConnection() {
+        return bootNodes.isEmpty() && pendingPingRequests.isEmpty() &&
+                pendingFindNodeRequests.isEmpty() && establishedConnections.isEmpty();
     }
```
[the remark this answers is not in the document]
Comment (casiojapi) [reply rskj-core/src/main/java/co/rsk/net/discovery/PeerExplorer.java:451]: I thought it was redundant but tbh it makes sense to firstly get closest nodes and then check them. thx

File: rskj-core/src/main/java/co/rsk/net/discovery/PeerExplorer.java [hunk 2]
```diff
@@ -427,12 +429,25 @@ synchronized void update() {
...
         }
 
-        List<Node> closestNodes = this.distanceTable.getClosestNodes(this.localNode.getId());
-
-        logger.trace("update - closestNodes: [{}]", closestNodes);
+        if (shouldRetryConnection()) {
+            retryCounter++;
+            if (retryCounter <= MAX_RETRY_ATTEMPTS) {
+                logger.info("retrying connection to bootstrap nodes. Attempt: {}", retryCounter);
+                startConversationWithNewNodes();
```
[the remark this answers is not in the document]
Comment (casiojapi) [reply rskj-core/src/main/java/co/rsk/net/discovery/PeerExplorer.java:436]: I thought the initialBootNodes loading done on the PeerExplorer constructor was enough:

`public PeerExplorer(List<String> initialBootNodes,
                        Node localNode, NodeDistanceTable distanceTable, ECKey key,
                        long reqTimeOut, long updatePeriod, long cleanPeriod, Integer networkId,
                        PeerScoringManager peerScoringManager, boolean allowMultipleConnectionsPerHostPort) {
        this.localNode = localNode;
        this.key = key;
        this.distanceTable = distanceTable;
        this.updateEntryLock = new ReentrantLock();
        this.networkId = networkId;
        loadInitialBootNodes(initialBootNodes);`
        
I'll test it with the loading inside the retry too. That's useful

File: rskj-core/src/main/java/co/rsk/net/discovery/PeerExplorer.java [hunk 3]
```diff
@@ -427,12 +429,25 @@ synchronized void update() {
             return;
         }
 
-        List<Node> closestNodes = this.distanceTable.getClosestNodes(this.localNode.getId());
-
-        logger.trace("update - closestNodes: [{}]", closestNodes);
+        if (shouldRetryConnection()) {
+            retryCounter++;
```
[the remark this answers is not in the document]
Comment (casiojapi) [reply rskj-core/src/main/java/co/rsk/net/discovery/PeerExplorer.java:433]: Yep, I was thinking of where to put the limit (wasn't sure about letting it go infinitely) 

I'm finishing the tests on the implementation. As soon as they pass I'll push the changes

File: rskj-core/src/main/resources/reference.conf [hunk 4]
```diff
@@ -142,6 +142,7 @@ peer {
     discovery = {
         # allow multiple connections per host by default
         allowMultipleConnectionsPerHostPort = true
+        maxBootRetries = -1
```
Comment (Vovchyk) [inline rskj-core/src/main/resources/reference.conf:145]:
```suggestion
        # allows to specify a number of attempts to discover at least one peer. By default, it's -1, which means an infinite number of attempts
        maxBootRetries = -1
```

File: rskj-core/src/main/java/co/rsk/net/discovery/PeerExplorer.java [hunk 5]
```diff
@@ -98,12 +101,13 @@ public class PeerExplorer {
...
                         Node localNode, NodeDistanceTable distanceTable, ECKey key,
                         long reqTimeOut, long updatePeriod, long cleanPeriod, Integer networkId,
-                        PeerScoringManager peerScoringManager, boolean allowMultipleConnectionsPerHostPort) {
+                        PeerScoringManager peerScoringManager, boolean allowMultipleConnectionsPerHostPort, long maxBootRetries) {
         this.localNode = localNode;
         this.key = key;
         this.distanceTable = distanceTable;
         this.updateEntryLock = new ReentrantLock();
         this.networkId = networkId;
+        this.initialBootNodes = initialBootNodes;
```
Comment (Vovchyk) [inline rskj-core/src/main/java/co/rsk/net/discovery/PeerExplorer.java:110]:
```suggestion
        this.initialBootNodes = Collections.unmodifiableList(new ArrayList<>(initialBootNodes));
```
minor thing, just to be on a safer side

**Review by Vovchyk — DISMISSED** [review]

File: rskj-core/src/main/java/co/rsk/net/discovery/PeerExplorer.java (hunk 5 above)
Comment (casiojapi) [reply rskj-core/src/main/java/co/rsk/net/discovery/PeerExplorer.java:110]: just added all the changes, running some tests to check everything it's ok

**Review by Vovchyk — APPROVED** [review]

**Review by Vovchyk — APPROVED** [review]

**Pull request merged**